### PR TITLE
add built-in policies for both subscriptions and management-groups

### PR DESCRIPTION
**Pull request by lawrenae**
this is meant as a fix for #2757 

Feedback most welcome, of course

### CODE REVIEW

**Comment by katbyte** [comment]
@lawrenae, should we expose more of the properties as the builtin role definition datasource does?

**Comment by lawrenae** [comment]
@katbyte sure. I don't know how useful those extra properties would be to folks, but it should be easy to add them. Will see what I can do!

**Comment by lawrenae** [comment]
@katbyte I added 
* `display_name` - (Required) Specifies the name of the built-in Policy Definition.
* `id` - the ID of the built-in Policy Definition.
* `description` - the Description of the built-in Policy.
* `type` - the Type of Policy.
* `name` - The Name of the Policy Definition

I see a few things in the portal that I couldnt find in the API -- `effect` and `category` that might also be useful.

File: azurerm/data_source_builtin_policy_definition.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+				Optional: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (katbyte) [inline azurerm/data_source_builtin_policy_definition.go]: Other properties i think we should export after looking at the resource are:
- `policy_type`
- `policy_rule`
- `metadata`
- `parameters`

File: azurerm/data_source_builtin_policy_definition.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+			}
+		}
+	}
+
+	if policyDefinition.ID == nil {
+		return fmt.Errorf("Error loading Policy Definition List: could not find policy '%s'", name)
+	}
+
+	d.SetId(*policyDefinition.ID)
+	d.Set("name", *policyDefinition.Name)
```
Comment (katbyte) [inline azurerm/data_source_builtin_policy_definition.go]: We don't have to dereference these here, `d.Set` will do it for us.

```suggestion
	d.Set("name", policyDefinition.Name)
```

File: website/docs/d/builtin_policy_definition.markdown [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_builtin_policy_definition"
+sidebar_current: "docs-azurerm-datasource-builtin-policy-definition"
+description: |-
+  Get information about an existing built-in Policy Definition.
+---
+
+# Data Source: azurerm_builtin_policy_definition
+
+Use this data source to access information about a built-in Policy Definition. To access information about a custom Policy Definition, [please see the `azurerm_policy_definition` data source](../r/policy_definition.html) instead.
```
Comment (katbyte) [inline website/docs/d/builtin_policy_definition.markdown]: The policy data source doesn't exist.. yet

File: azurerm/data_source_builtin_policy_definition.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (katbyte) [inline azurerm/data_source_builtin_policy_definition.go]: This should be computed as we are only returning the value here

**Comment by lawrenae** [comment]
@katbyte I like the idea of combining them. I will get to working on that, hopefully very soon

File: azurerm/data_source_policy_definition.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+func dataSourceArmPolicyDefinition() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceArmPolicyDefinitionRead,
+		Schema: map[string]*schema.Schema{
+			"display_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: Could we validate this with `validate.NoZerovalues`?

File: azurerm/data_source_policy_definition.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+			"management_group_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: The could be validate with `azure.VAlidateresourceIDOrEmpty`

File: azurerm/data_source_policy_definition.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: As this is a data source this should be computed not optional
```suggestion
				Computed: true,
```

File: azurerm/data_source_policy_definition.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: As this is a data source this should be computed not optional
```suggestion
				Computed: true,
```

File: azurerm/data_source_policy_definition.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: As this is a data source this should be computed not optional
```suggestion
				Computed: true,
```

File: azurerm/data_source_policy_definition.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+	if err != nil {
+		return fmt.Errorf("Error loading Policy Definition List: %+v", err)
+	}
+
+	var policyDefinition policy.Definition
+
+	for policyDefinitions.NotDone() {
+		def := policyDefinitions.Value()
+		if *def.DisplayName == name {
```
Comment (katbyte) [inline azurerm/data_source_policy_definition.go]: You may want to check if `def.DisplayName` is ` nil` before dereferencing it

File: azurerm/data_source_policy_definition_test.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func testAccDataSourceBuiltInPolicyDefinitionAtManagementGroup(name string) string {
+	return fmt.Sprintf(`
+
+data "azurerm_client_config" "current" {}
+
+data "azurerm_policy_definition" "test" {
+  display_name = "%s"
+  management_group_id = "${data.azurerm_client_config.current.tenant_id}"
```
Comment (katbyte) [inline azurerm/data_source_policy_definition_test.go]: Could we align these assignments?

File: azurerm/data_source_policy_definition.go (hunk 5 above)
Comment (lawrenae) [reply azurerm/data_source_policy_definition.go]: I went with `validate.NoEmptyStrings` instead -- cool?

**Comment by lawrenae** [comment]
@katbyte I appreciate the additional feedback, and think I've got this ready for review (again).

**Comment by lawrenae** [comment]
@katbyte -- good catches on the docs. updated!

**Comment by katbyte** [comment]
Thanks @lawrenae, LGTM now 🙂

**Pull request merged**